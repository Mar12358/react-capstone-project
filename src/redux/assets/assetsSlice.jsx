import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const urlAssets = 'https://api.coincap.io/v2/assets';

const initialState = {
  assets: [],
  isLoading: false,
  error: undefined,
};

export const getAssets = createAsyncThunk(
  'Assets/getAssets',
  async (thunkAPI) => {
    try {
      const resp = await axios(urlAssets);
      return resp.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error);
    }
  },
);

const assetsSlice = createSlice({
  name: 'assets',
  initialState,
  reducers: {
    /* searchAssetByName: (state, action) => {
      const newState = state;
      const assetName = action.payload;
      console.log(assetName);
      // newState.assets = state.assets.filter((asset) => asset.name === assetName);
    }, */
    /* cancelReserveRocket: (state, action) => {
      const rocketId = action.payload;
      const newState = state;
      newState.rockets = state.rockets.map((rocket) => (
        rocket.rocket_id !== rocketId ? rocket : { ...rocket, reserved: false }
      ));
    }, */
  },
  extraReducers: (builder) => {
    builder
      .addCase(getAssets.pending, (state) => {
        const pendingState = state;
        pendingState.isLoading = true;
      })
      .addCase(getAssets.fulfilled, (state, action) => {
        const fulfilledState = state;
        fulfilledState.isLoading = false;
        fulfilledState.assets = action.payload;
      })
      .addCase(getAssets.rejected, (state) => {
        const rejectedState = state;
        rejectedState.isLoading = false;
      });
  },
});

export const { searchAssetByName } = assetsSlice.actions;
export default assetsSlice.reducer;
