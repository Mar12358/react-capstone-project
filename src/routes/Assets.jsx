import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
// import Button from 'react-bootstrap/Button';
// import Badge from 'react-bootstrap/Badge';
import { useHistory } from 'react-router-dom';
import {
  getAssets,
} from '../redux/assets/assetsSlice';
import styles from '../styles/assets.module.css';

const Assets = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const { assets, isLoading, error } = useSelector((state) => state.assets);
  console.log('assets', assets.data);
  useEffect(() => {
    if (assets.length === 0) {
      dispatch(getAssets());
    }
  }, [dispatch, assets.length]);

  const handleAssetClick = (symbolId) => {
    history.push(`/assets/${symbolId}`);
  };

  /* const handleCancelReserveRocket = (id) => {
    dispatch(cancelReserveRocket(id));
  }; */
  if (error) return <h2>Something went wrong</h2>;
  let cont = 0;
  let aux = 0;
  let switching = false;
  return (isLoading || assets.length === 0) ? (
    <h1>Loading...</h1>
  ) : (
    <>
      <hr />
      <div className={styles.assetsWrapper}>
        {assets.data.map((asset) => {
          cont += 1;
          if (cont - aux === 3) { switching = !switching; aux = cont - 1; }
          return switching ? (
            <button
              key={asset.id}
              className={(cont % 2 === 0) ? styles.assetB : styles.assetA}
              type="button"
              onClick={() => handleAssetClick(asset.id)}
            >
              <div className={styles.symbol}>{asset.symbol}</div>
              <div className={styles.smallInfo}>
                <p className={styles.name}>{asset.name}</p>
                <p className={styles.priceUsd}>
                  U$D:
                  {' '}
                  {parseFloat(asset.priceUsd).toFixed(4)}
                </p>
              </div>
            </button>
          ) : (
            <button
              key={asset.id}
              className={(cont % 2 === 0) ? styles.assetA : styles.assetB}
              type="button"
              onClick={() => handleAssetClick(asset.id)}
            >
              <div className={styles.symbol}>{asset.symbol}</div>
              <div className={styles.smallInfo}>
                <p className={styles.name}>{asset.name}</p>
                <p className={styles.priceUsd}>
                  U$D:
                  {' '}
                  {parseFloat(asset.priceUsd).toFixed(4)}
                </p>
              </div>
            </button>
          );
        })}
      </div>
    </>
  );
};
export default Assets;
