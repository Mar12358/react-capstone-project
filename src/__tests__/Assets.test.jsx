import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import Assets from '../routes/Assets';
import { getAssets } from '../redux/assets/assetsSlice';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.mock('../redux/assets/assetsSlice', () => ({
  getAssets: jest.fn(),
}));

describe('Assets', () => {
  const mockDispatch = jest.fn();
  const mockNavigate = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(mockDispatch);
    useSelector.mockReturnValue({
      filterData: [
        {
          id: 1, symbol: 'BTC', name: 'Bitcoin', priceUsd: '50000',
        },
        {
          id: 2, symbol: 'ETH', name: 'Ethereum', priceUsd: '2000',
        },
      ],
      isLoading: false,
      error: undefined,
    });
    useNavigate.mockReturnValue(mockNavigate);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders assets correctly', () => {
    render(<Assets />);

    expect(screen.getByText('BTC')).toBeInTheDocument();
    expect(screen.getByText('Bitcoin')).toBeInTheDocument();
    expect(screen.getByText('U$D: 50000.0000')).toBeInTheDocument();

    expect(screen.getByText('ETH')).toBeInTheDocument();
    expect(screen.getByText('Ethereum')).toBeInTheDocument();
    expect(screen.getByText('U$D: 2000.0000')).toBeInTheDocument();

    expect(screen.getByText('BTC')).toMatchSnapshot();
    expect(screen.getByText('ETH')).toMatchSnapshot();
  });

  test('dispatches getAssets action if filterData is empty', () => {
    useSelector.mockReturnValueOnce({
      filterData: [],
      isLoading: false,
      error: undefined,
    });

    render(<Assets />);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(getAssets).toHaveBeenCalled();
  });

  test('navigates to asset details on button click', () => {
    render(<Assets />);

    const btcButton = screen.getByText('BTC');
    const ethButton = screen.getByText('ETH');

    btcButton.click();
    expect(mockNavigate).toHaveBeenCalledWith('/assets/1');

    ethButton.click();
    expect(mockNavigate).toHaveBeenCalledWith('/assets/2');
  });

  test('renders loading state', () => {
    useSelector.mockReturnValueOnce({
      filterData: [],
      isLoading: true,
      error: undefined,
    });

    render(<Assets />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(screen.getByText('Loading...')).toMatchSnapshot();
  });

  test('renders error message', () => {
    useSelector.mockReturnValueOnce({
      filterData: [],
      isLoading: false,
      error: 'Something went wrong',
    });

    render(<Assets />);

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    expect(screen.getByText('Something went wrong')).toMatchSnapshot();
  });
});
