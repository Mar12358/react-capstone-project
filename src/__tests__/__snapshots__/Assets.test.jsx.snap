// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets renders assets correctly 1`] = `
<div
  class="symbol"
>
  BTC
</div>
`;

exports[`Assets renders assets correctly 2`] = `
<div
  class="symbol"
>
  ETH
</div>
`;

exports[`Assets renders error message 1`] = `
<h2>
  Something went wrong
</h2>
`;

exports[`Assets renders loading state 1`] = `
<h1>
  Loading...
</h1>
`;
